File: windows/UniviewCodeBuddyProxy.Server/Services/CacheManager.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace UniviewCodeBuddyProxy.Services;

/// <summary>
/// Thread-safe TTL memory cache with SHA256 key generation.
/// </summary>
public sealed class CacheManager
{
    // ═══ Constants ═══

    private const int CleanupThreshold = 10_000;
    public const int DefaultTTL = 300;

    // ═══ State ═══

    private readonly Dictionary<string, CacheEntry> _store = new();
    private TimeSpan _ttl = TimeSpan.FromSeconds(DefaultTTL);
    private bool _enabled;
    private bool _isCleaningUp;
    private readonly object _lock = new();

    // ═══ Enable/Disable ═══

    public void SetEnabled(bool enabled)
    {
        lock (_lock)
        {
            _enabled = enabled;
            if (!enabled)
                _store.Clear();
        }
    }

    public bool IsEnabled
    {
        get { lock (_lock) { return _enabled; } }
    }

    public void SetTTL(int ttlSeconds)
    {
        lock (_lock) { _ttl = TimeSpan.FromSeconds(ttlSeconds); }
    }

    // ═══ Cache Key ═══

    /// <summary>
    /// Build a cache key from model, messages, tools, temperature, and maxTokens using SHA256.
    /// </summary>
    public static string BuildKey(string model, object? messages, object? tools, double temperature, int maxTokens)
    {
        using var sha = SHA256.Create();
        var modelBytes = Encoding.UTF8.GetBytes(model);
        sha.TransformBlock(modelBytes, 0, modelBytes.Length, null, 0);

        if (messages != null)
        {
            var json = JsonSerializer.Serialize(messages);
            var bytes = Encoding.UTF8.GetBytes(json);
            sha.TransformBlock(bytes, 0, bytes.Length, null, 0);
        }

        if (tools != null)
        {
            var json = JsonSerializer.Serialize(tools);
            var bytes = Encoding.UTF8.GetBytes(json);
            sha.TransformBlock(bytes, 0, bytes.Length, null, 0);
        }

        var tempBytes = Encoding.UTF8.GetBytes(temperature.ToString("R"));
        sha.TransformBlock(tempBytes, 0, tempBytes.Length, null, 0);

        var maxBytes = Encoding.UTF8.GetBytes(maxTokens.ToString());
        sha.TransformBlock(maxBytes, 0, maxBytes.Length, null, 0);

        sha.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
        return Convert.ToHexString(sha.Hash!).ToLowerInvariant();
    }

    // ═══ Get/Set ═══

    /// <summary>Get a cache entry, returns null if not found or expired.</summary>
    public byte[]? Get(string key)
    {
        lock (_lock)
        {
            if (!_enabled) return null;

            if (!_store.TryGetValue(key, out var entry))
                return null;

            if (DateTime.UtcNow >= entry.Expiration)
            {
                TriggerCleanup();
                return null;
            }

            return entry.Data;
        }
    }

    /// <summary>Store a cache entry.</summary>
    public void Set(string key, byte[] value, int? ttlSeconds = null)
    {
        lock (_lock)
        {
            if (!_enabled) return;

            var effectiveTTL = ttlSeconds.HasValue
                ? TimeSpan.FromSeconds(ttlSeconds.Value)
                : _ttl;

            _store[key] = new CacheEntry(value, DateTime.UtcNow + effectiveTTL);

            if (_store.Count > CleanupThreshold)
                TriggerCleanup();
        }
    }

    // ═══ Invalidate/Clear ═══

    /// <summary>Invalidate cache entries matching a prefix.</summary>
    public void Invalidate(string prefix)
    {
        lock (_lock)
        {
            var keysToRemove = new List<string>();
            foreach (var key in _store.Keys)
            {
                if (key.StartsWith(prefix))
                    keysToRemove.Add(key);
            }
            foreach (var key in keysToRemove)
                _store.Remove(key);
        }
    }

    /// <summary>Clear all cache entries.</summary>
    public void Clear()
    {
        lock (_lock)
        {
            _store.Clear();
        }
    }

    // ═══ Cleanup ═══

    private void TriggerCleanup()
    {
        if (_isCleaningUp) return;
        _isCleaningUp = true;

        try
        {
            var now = DateTime.UtcNow;
            var keysToRemove = new List<string>();
            foreach (var (key, entry) in _store)
            {
                if (now >= entry.Expiration)
                    keysToRemove.Add(key);
            }
            foreach (var key in keysToRemove)
                _store.Remove(key);
        }
        finally
        {
            _isCleaningUp = false;
        }
    }

    // ═══ Entry ═══

    private sealed record CacheEntry(byte[] Data, DateTime Expiration);
}

File: macos/UniviewCodeBuddyProxy/Sources/Services/CacheManager.swift
import CryptoKit
import Foundation

// ═══════════════════════════════════════════════
// CacheManager — 基于 actor 的内存 TTL 响应缓存
// ═══════════════════════════════════════════════

/// 缓存条目
private struct CacheEntry {
	let data: Data
	let expiration: Date
}

/// 基于 actor 的线程安全 TTL 内存缓存
actor CacheManager {

	// MARK: - 常量

	/// 默认 TTL（秒）
	static let defaultTTL: TimeInterval = 300

	/// 触发惰性清理的条目数阈值
	private static let cleanupThreshold = 10_000

	// MARK: - 属性

	private var store: [String: CacheEntry] = [:]
	private var ttl: TimeInterval
	private var enabled: Bool = false
	private var isCleaningUp: Bool = false

	// MARK: - Init

	init(ttl: TimeInterval = CacheManager.defaultTTL) {
		self.ttl = ttl
	}

	// MARK: - 启用/禁用

	/// 启用或禁用缓存
	func setEnabled(_ enabled: Bool) {
		self.enabled = enabled
		if enabled && store.isEmpty {
			store = [:]
		}
	}

	/// 缓存是否启用
	func isEnabled() -> Bool {
		enabled
	}

	/// 设置缓存 TTL
	func setTTL(_ ttl: TimeInterval) {
		self.ttl = ttl
	}

	// MARK: - 缓存 Key 构建

	/// 根据 model、messages、tools、temperature、maxTokens 构建缓存 key
	/// 使用 SHA-256 哈希保证 key 长度固定且唯一
	static func key(
		model: String,
		messages: Any?,
		tools: Any?,
		temperature: Double,
		maxTokens: Int
	) -> String {
		var hasher = SHA256()
		hasher.update(data: Data(model.utf8))

		if let messages = messages {
			if let data = try? JSONSerialization.data(withJSONObject: messages) {
				hasher.update(data: data)
			}
		}

		if let tools = tools {
			if let data = try? JSONSerialization.data(withJSONObject: tools) {
				hasher.update(data: data)
			}
		}

		hasher.update(data: Data(String(temperature).utf8))
		hasher.update(data: Data(String(maxTokens).utf8))

		let digest = hasher.finalize()
		return digest.compactMap { String(format: "%02x", $0) }.joined()
	}

	// MARK: - Get / Set

	/// 获取缓存条目，过期返回 nil
	func get(key: String) -> Data? {
		guard enabled else { return nil }

		guard let entry = store[key] else {
			return nil
		}

		guard Date() < entry.expiration else {
			// 过期：异步触发清理（actor 内部直接调用即可）
			triggerCleanup()
			return nil
		}

		return entry.data
	}

	/// 存储缓存条目
	func set(key: String, value: Data, ttl: TimeInterval? = nil) {
		guard enabled else { return }

		let effectiveTTL = ttl ?? self.ttl
		let expiration = Date().addingTimeInterval(effectiveTTL)

		store[key] = CacheEntry(data: value, expiration: expiration)

		// 惰性清理：条目数超过阈值时触发
		if store.count > Self.cleanupThreshold {
			triggerCleanup()
		}
	}

	// MARK: - 失效/清空

	/// 使匹配前缀的缓存条目失效
	func invalidate(prefix: String) {
		store = store.filter { !$0.key.hasPrefix(prefix) }
	}

	/// 清空所有缓存
	func clear() {
		store = [:]
	}

	// MARK: - 清理

	/// 触发过期条目清理（防止并发 cleanup）
	private func triggerCleanup() {
		guard !isCleaningUp else { return }
		isCleaningUp = true
		cleanup()
		isCleaningUp = false
	}

	/// 删除所有过期条目
	private func cleanup() {
		let now = Date()
		store = store.filter { now < $0.value.expiration }
	}
}
